spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -0,0 +1,165 @@
+            Alias(monotonically_increasing_id().expr, ROW_ID_COL)()
+          )
+        val newTargetTable = Project(newProjectCols, target)
+
+        // Construct the plan to prune target based on join condition between source and
+        // target.
+        val prunedTargetPlan = Join(source, newTargetTable, Inner, Some(cond), JoinHint.NONE)
+        val writeInfo = newWriteInfo(target.schema)
+        val mergeBuilder = target.table.asMergeable.newMergeBuilder("delete", writeInfo)
+        val targetTableScan =  buildScanPlan(spark, target.table,
+          target.output, mergeBuilder, None, Some(prunedTargetPlan),
+          isCountCheckEnabled(target.table))
+
+        // Construct an outer join to help track changes in source and target.
+        // TODO : Optimize this to use LEFT ANTI or RIGHT OUTER when applicable.
+        val sourceTableProj = source.output ++ Seq(Alias(lit(true).expr, ROW_FROM_SOURCE)())
+        val targetTableProj = target.output ++ Seq(Alias(lit(true).expr, ROW_FROM_TARGET)())
+        val newTargetTableScan = Project(targetTableProj, targetTableScan)
+        val newSourceTableScan = Project(sourceTableProj, source)
+        val joinPlan = Join(newSourceTableScan, newTargetTableScan, FullOuter, Some(cond), JoinHint.NONE)
+
+        // Construct the plan to replace the data based on the output of `MergeInto`
+        val mergeParams = MergeIntoParams(
+          isSourceRowNotPresent = IsNull(findOutputAttr(joinPlan, ROW_FROM_SOURCE)),
+          isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan, ROW_FROM_TARGET)),
+          matchedConditions = actions.map(getClauseCondition),
+          matchedOutputs = actions.map(actionOutput(_, targetOutputCols)),
+          notMatchedConditions = notActions.map(getClauseCondition),
+          notMatchedOutputs = notActions.map(actionOutput(_, targetOutputCols)),
+          targetOutput = targetOutputCols :+ Literal(false),
+          deleteOutput = targetOutputCols :+ Literal(true),
+          joinedAttributes = joinPlan.output
+        )
+        val joinKeysFromTarget = targetOutputCols.filter (
+          attr => cond.references.exists(attr.semanticEquals(_))
+        )
+        val writePlan = buildWritePlan(joinPlan, target.table, joinKeysFromTarget)
+        val mergePlan = MergeInto(mergeParams, target, writePlan)
+        val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+        ReplaceData(target, batchWrite, mergePlan)
+    }
+  }
+
+  def buildWritePlan(plan: LogicalPlan, table: Table, targetJoinAttrs: Seq[Attribute]): LogicalPlan = {
+     val iceTable = table.asInstanceOf[SparkTable].table()
+     val globalSortEnabled = isGlobalSortEnabled(table)
+     val partitionExpressions = toCatalyst(iceTable.spec(), plan)
+     val sortExpressions =  toCatalyst(iceTable.sortOrder(), plan, iceTable.schema())
+    val numShufflePartitions = SQLConf.get.numShufflePartitions
+    (partitionExpressions.isEmpty, sortExpressions.isEmpty) match {
```

Comment:
I don't think the logic here that always adds `RepartitionByExpression` is correct. If the sort is global, that will automatically add the repartition by expression required by that sort, right? And that might not be the same as the partition expressions. When performing a global sort, the partition expressions and sort must be merged to produce a sort that meets the sort requirements within each partition.